File: third-semester/homework1.1/LazyEvaluation/LazyFactory.cs
namespace LazyEvaluation
{
    using System;

    /// <summary>
    /// Class for creating lazy evaluation objects
    /// </summary>
    public static class LazyFactory
    {
        /// <summary>
        /// Creates new instance of simple lazy evaluation implementation
        /// </summary>
        /// <param name="supplier">function that represents evaluation</param>
        /// <returns>simple lazy evaluation object</returns>
        /// <typeparam name="T">type of evaluation result</typeparam>
        public static ILazy<T> CreateSingleThreadedLazy<T>(Func<T> supplier)
            => new SingleThreadedLazy<T>(supplier);

        /// <summary>
        /// Creates new instance of thread-safe lazy evaluation implementation
        /// </summary>
        /// <param name="supplier">function that represents evaluation</param>
        /// <returns>thread-safe lazy evaluation object</returns>
        /// <typeparam name="T">type of evaluation result</typeparam>
        public static ILazy<T> CreateMultiThreadedLazy<T>(Func<T> supplier)
            => new MultiThreadedLazy<T>(supplier);
    }
}

File: 4th-semester/Homework8.1/Lazy/LazyFactory.fs
namespace Lazy

/// Lazy object factory
type LazyFactory() =
    /// Creates single threaded lazy object
    static member CreateSingleThreadedLazy<'a>(supplier: unit -> 'a) =
        SingleThreadedLazy supplier :> ILazy<'a>
        
    static member CreateMultiThreadedLazy<'a>(supplier: unit -> 'a) =
        MultiThreadedLazy supplier :> ILazy<'a>
        
    static member CreateLockfreeLazy<'a>(supplier: unit -> 'a) =
        LockfreeLazy supplier :> ILazy<'a>